c/partial_composable_lock.c

```diff
@@ -0,0 +1,322 @@
+  }
+  const uint8_t *current_identity = &args_bytes_seg.ptr[33];
+  size_t identity_size = args_bytes_seg.size - 33;
+  if (identity_size > IDENTITY_BUFFER_SIZE) {
+    DEBUG("Identity is too large!");
+    return ERROR_TRANSACTION;
+  }
+
+  uint8_t code_buffer[CODE_SIZE] __attribute__((aligned(RISCV_PGSIZE)));
+  uint8_t prefilled_data_buffer[PREFILLED_DATA_SIZE];
+  CkbSwappableSignatureInstance instance;
+  instance.code_buffer = code_buffer;
+  instance.code_buffer_size = CODE_SIZE;
+  instance.prefilled_data_buffer = prefilled_data_buffer;
+  instance.prefilled_buffer_size = PREFILLED_DATA_SIZE;
+  instance.verify_func = NULL;
+
+  ret = ckb_initialize_swappable_signature(args_bytes_seg.ptr,
+                                           args_bytes_seg.ptr[32], &instance);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  size_t input_i = 0, output_i = 0;
+  uint8_t witness[SIGNATURE_WITNESS_BUFFER_SIZE];
+  len = SIGNATURE_WITNESS_BUFFER_SIZE;
+  // Initial witness load
+  ret = ckb_load_witness(witness, &len, 0, input_i, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  size_t readed_len = len;
+  if (readed_len > SIGNATURE_WITNESS_BUFFER_SIZE) {
+    readed_len = SIGNATURE_WITNESS_BUFFER_SIZE;
+  }
+
+  // Loop through each partial transaction
+  int has_more = 1;
+  while (has_more) {
+    // The first witness in each partial transaction must contain a valid
+    // signature.
+    size_t start_input_index = input_i;
+    uint8_t signature[SIGNATURE_BUFFER_SIZE];
+    size_t signature_length;
+    uint8_t output_count;
+
+    if (readed_len < 20) {
+      DEBUG("Invalid signature witness length!");
+      return ERROR_TRANSACTION;
+    }
+    uint32_t lock_length = *((uint32_t *)(&witness[16]));
+    if (readed_len < 20 + lock_length) {
+      DEBUG("Witness lock part is far tooooo long!");
+      return ERROR_TRANSACTION;
+    }
+    if (lock_length <= 1) {
+      DEBUG("Signature is required!");
+      return ERROR_TRANSACTION;
+    }
+    if (lock_length - 1 > SIGNATURE_BUFFER_SIZE) {
+      DEBUG("Signature is too long!");
+      return ERROR_TRANSACTION;
+    }
+    uint8_t *lock = &witness[20];
+    // To ensure the security of a partial transaction, a script hash is always
+    // included in the witness. There must be one running script in the current
+    // transaction with matching script hash here.
+    ret = script_exists(lock);
+    if (ret != CKB_SUCCESS) {
+      return ret;
+    }
+    output_count = lock[32];
+    signature_length = lock_length - 33;
+    memcpy(signature, &lock[33], signature_length);
+    memset(&lock[33], 0, signature_length);
+
+    blake2b_state message_ctx;
+    blake2b_init(&message_ctx, BLAKE2B_BLOCK_SIZE);
+    // Hash the first witness
+    blake2b_update(&message_ctx, witness, readed_len);
+    if (readed_len < len) {
+      ret = ckb_load_and_hash(&message_ctx, readed_len, input_i,
+                              CKB_SOURCE_GROUP_INPUT, ckb_load_witness);
```

Comment:
First witness already updated into `message_ctx`, since `ckb_load_and_hash ` will include witness at [`index`](https://github.com/nervosnetwork/ckb-c-stdlib/blob/7cb529ada75e93204650cba95681144f06fedd19/ckb_streaming.h#L17-L23),  so the 3rd argument should be `input_i + 1`?